framework/wcf/Extensibility/Transport/AzureQueueStorageClient/AzureQueueStorageClient/AzureQueueStorageOutputChannel.cs

```diff
@@ -0,0 +1,263 @@
+            {
+                return (T)(object)this;
+            }
+
+            T messageEncoderProperty = this.encoder.GetProperty<T>();
+            if (messageEncoderProperty != null)
+            {
+                return messageEncoderProperty;
+            }
+
+            return base.GetProperty<T>();
+        }
+
+        /// <summary>
+        /// Open the channel for use. We do not have any blocking work to perform so this is a no-op
+        /// </summary>
+        protected override void OnOpen(TimeSpan timeout)
+        {
+        }
+
+        protected override IAsyncResult OnBeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return new CompletedAsyncResult(callback, state);
+        }
+
+        protected override void OnEndOpen(IAsyncResult result)
+        {
+            CompletedAsyncResult.End(result);
+        }
+
+
+        #region Socket_Shutdown
+        /// <summary>
+        /// Shutdown ungracefully
+        /// </summary>
+        protected override void OnAbort()
+        {
+
+        }
+
+        /// <summary>
+        /// Shutdown gracefully
+        /// </summary>
+        protected override void OnClose(TimeSpan timeout)
+        {
+
+        }
+
+        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            this.OnClose(timeout);
+            return new CompletedAsyncResult(callback, state);
+        }
+
+        protected override void OnEndClose(IAsyncResult result)
+        {
+            CompletedAsyncResult.End(result);
+        }
+        #endregion
+
+        #region Send_Synchronous
+        /// <summary>
+        /// Address the Message and serialize it into a byte array.
+        /// </summary>
+        ArraySegment<byte> EncodeMessage(Message message)
+        {
+            try
+            {
+                this.remoteAddress.ApplyTo(message);
+                return encoder.WriteMessage(message, int.MaxValue, parent.BufferManager);
+            }
+            finally
+            {
+                // We have consumed the message by serializing it, so clean up
+                message.Close();
+            }
+        }
+
+        public void Send(Message message)
+        {
+            try
+            {
+                queueClient.SendMessage(message.ToString());
+            }
+            catch
+            {
+
+            }
+            finally
+            {
+               
+            }
+        }
+
+        public void Send(Message message, TimeSpan timeout)
+        {
+            // UDP does not block so we do not need timeouts.
+            this.Send(message);
+        }
+        #endregion
+
+        #region Send_Asynchronous
+        public IAsyncResult BeginSend(Message message, AsyncCallback callback, object state)
+        {
+            //base.ThrowIfDisposedOrNotOpen();
+            return new SendAsyncResult(this, message, callback, state);
+        }
```

Comment:
You are completely missing the async implementation in SendAsyncResult. Thankfully it's a lot easier to write this code now. Take a look in the dotnet/wcf codebase for a class called TaskHelpers. You need to copy over 2 methods, ToApm and ToApmEnd (there's generic (non-void return type) and non generic (used when returning void) variants. I think you only need the non-generic variant. Then add a private method with the signature `private async Task SendAsync(Message message, TimeSpan timeout)` to do the work. This method should look the same as the synchronous Send method, but call the async variants of methods and await them. Don't worry about the async variation of message encoding, it's only useful when streaming data. I think you will only use the async queueClient.SendMessageAsync variant. You will need to create a CancellationTokenSource from the passed in timeout value, and dispose it in the finally (to cancel the timer). Don't forget to use ConfigureAwait(false) on the SendMessageAsync method.

Reply:
Thank you for the tips. Updated